File: languages/js/src/Tracer.js
import { maxTracers, maxTraces } from '../../../limits';

class Tracer {
  static serialize(object) {
    return JSON.parse(JSON.stringify(object));
  }

  static addTracer(className, title) {
    const key = `${this.tracerCount++}-${className}-${title}`;
    const method = 'construct';
    const args = [className, title];
    this.addTrace(key, method, args);
    return key;
  }

  static addTrace(tracerKey, method, args) {
    const trace = {
      tracerKey,
      method,
      args: this.serialize(args.map(arg => arg instanceof Tracer ? arg.key : arg)),
    };
    this.traces.push(trace);
    if (this.traces.length > maxTraces) throw new Error('Traces Limit Exceeded');
    if (this.tracerCount > maxTracers) throw new Error('Tracers Limit Exceeded');
  }

  constructor(title = this.constructor.name) {
    this.key = Tracer.addTracer(this.constructor.name, title);
    this.register(
      'reset',
      'set',
      'delay',
    );
  }

  register(...functions) {
    for (const func of functions) {
      this[func] = (...args) => {
        Tracer.addTrace(this.key, func, args);
        return this;
      };
    }
  }

  unregister(...functions) {
    for (const func of functions) {
      delete this[func];
    }
  }
}

Tracer.tracerCount = 0;
Tracer.traces = [];

export default Tracer;

File: languages/cpp/src/Tracer.cpp
#include <string>
#include <iostream>
#include "json.hpp"
#include "Tracer.h"

using namespace std;
using json = nlohmann::json;

int Tracer::tracerCount = 0;
json Tracer::traces = json::array();

string Tracer::addTracer(string className, string title) {
    string key = to_string(tracerCount++) + "-" + className + "-" + title;
    string method = "construct";
    addTrace(key, method, json::array({className, title}));
    return key;
}

void Tracer::addTrace(string tracerKey, string method, json args) {
    traces.push_back({
                             {"tracerKey", tracerKey},
                             {"method",    method},
                             {"args",      args},
                     });
    if (traces.size() > maxTraces) throw overflow_error("Traces Limit Exceeded");
    if (tracerCount > maxTracers) throw overflow_error("Tracers Limit Exceeded");
}

Tracer::Tracer(string className, string title) {
    if (title.empty()) title = className;
    key = addTracer(className, title);
}

void Tracer::onExit() {
    cout << traces.dump(2);
}

int init() {
    atexit(Tracer::onExit);
    return 0;
}

int _ = init();